src/lib/utils/object-graph.test.ts

```diff
@@ -0,0 +1,146 @@
+
+    // @ts-expect-error Child can't be defined yet; that's what the circular
+    // reference is all about
+    const parent: Parent = { name: 'parent' }
+    const child: Child = { name: 'child', parent }
+    parent.child = child
+
+    const result = flattenObjectGraph(parent)
+    const uuids = Object.keys(result.include)
+    expect(uuids.length).toBe(1)
+
+    const roundTrip = inflateObjectGraph(result)
+    expect(roundTrip.child.parent).toBe(roundTrip)
+  })
+
+  it('does not create includes for primitives', () => {
+    const input = { a: 1, b: true, c: 'text' }
+    const result = flattenObjectGraph(input)
+
+    expect(Object.keys(result.include)).toHaveLength(0)
+    expect(inflateObjectGraph(result)).toEqual(input)
+  })
+
+  it('preserves repeated references in arrays', () => {
+    const shared = { thing: 'yo' }
+    const input = [shared, shared]
+    const result = flattenObjectGraph(input)
+
+    expect(Object.keys(result.include)).toHaveLength(1)
+    expect(result.data[0]).toEqual(result.data[1])
+    const deserializedData = inflateObjectGraph(result)
+    expect(deserializedData[0]).toBe(deserializedData[1])
+  })
+
+  it('deduplicates shared leaf objects', () => {
+    const leaf = { id: 999 }
+    const input = {
+      left: { leaf },
+      right: { leaf },
+    }
+
+    const result = flattenObjectGraph(input)
+    expect(Object.keys(result.include)).toHaveLength(1)
+    const roundTrip = inflateObjectGraph(result)
+
+    expect(roundTrip.left.leaf).toBe(roundTrip.right.leaf)
+  })
+
+  it('handles complex shared and circular structures', () => {
+    const shared = { val: 123 }
+    const a: Record<string, unknown> = { name: 'A', shared }
+    const b: Record<string, unknown> = { name: 'B', shared }
+    a.b = b
+    b.a = a
+
+    const result = flattenObjectGraph(a)
+    const roundTrip = inflateObjectGraph(result)
+
+    // TS gets a little (reasonably) annoyed that we're not more specific with
+    // the static types. We know they're there.
+    //
+    // @ts-expect-error It's really there; trust me
+    expect(roundTrip.b.a).toBe(roundTrip)
+    // @ts-expect-error It's really there; trust me
+    expect(roundTrip.shared).toBe(roundTrip.b.shared)
+  })
+
+  it('handles circular references inside nested arrays', () => {
+    const node: Record<string, unknown> = { id: 1 }
+    node.children = [node]
+
+    const result = flattenObjectGraph(node)
+    const roundTrip = inflateObjectGraph(result)
+
+    expect(roundTrip.children[0]).toBe(roundTrip)
+  })
+
+  it('handles null and skips undefined', () => {
+    const input = {
+      a: null,
+      b: undefined,
+      c: { d: undefined, e: null },
+    }
+
+    const result = flattenObjectGraph(input)
+    const roundTrip = inflateObjectGraph(result)
+
+    expect(roundTrip.a).toBeNull()
+    expect(roundTrip.b).toBeUndefined()
+    expect(roundTrip.c.e).toBeNull()
+    expect('d' in roundTrip.c).toBe(true) // optional: assert 'undefined' is retained
```

Comment:
Should we be making the same two assertions for `b` and `d`? Test that `d` is undefined but that `b` is a key in the object?

And is "skips undefined" an accurate description of what's happening? From reading the tests, it seems that `{ b: undefined }` gets translated into exactly the same thing in the round-trip object (value of undefined but the key exists).

Reply:
Oh, y'know what? "Skips undefined" is 100% wrong. Good catch! If we tried to serialize the output of `flattenObjectGraph` into JSON when the input contained `undefined` values, the JSON would just drop the properties altogether. That's not what this does...and honestly, I think that's fine? 🤔

Unless you've got an argument to change the behavior to _actually_ drop properties whose values are `undefined`, I'll just change the description of the test.